### ISSUE TITLE
Assertion failed when casting function pointers. "This is a bug in the Zig compiler"

### ISSUE THREAD

**Issue by davi2205** [ziglang/zig#6280]
**Zig compiler version (installed using snap):**
```
0.6.0+a8a806e92
```
**Operating System:**
```
Ubuntu 20.04.1 LTS 
```
**Description:**

As I was playing with Zig, I attempted to cast function pointer types and somehow the compiler "crashed".

**The code:**

```zig
const out = @import("std").io.getStdOut().writer();

const CallbackEntry = struct {
    pub const Context = @Type(.Opaque);

    target: *Context,
    func: fn(*Context)void,
};

const Foo = struct {
    bar: i32,
    baz: i32,
};

fn fooCallback(foo: *Foo) void {
    out.print("Yo! I was just called. bar = {}, baz = {}.", .{foo.bar, foo.baz}) catch unreachable;
}

pub fn main() !void {
    var foo = Foo {
        .bar = 30,
        .baz = 50,
    };

    const entry = CallbackEntry {
        .target = @ptrCast(*CallbackEntry.Context, &foo),
        .func = @ptrCast(fn(*CallbackEntry.Context)void, &fooCallback), // it fails here.
    };

    entry.func.*(entry.target);
}
```
**Output:**
```
when analyzing /home/davi/test.zig:27:9 in compiler source at /deps/zig/src/ir.cpp:20004: assertion failed. This is a bug in the Zig compiler.
Unable to dump stack trace: debug info stripped
Aborted (core dumped)
```

**Comment by davi2205** [comment]
I found out what was wrong in my code
```zig
const out = @import("std").io.getStdOut().writer();

const CallbackEntry = struct {
    pub const Context = @Type(.Opaque);

    target: *Context,
    func: fn(*Context)void,
};

const Foo = struct {
    bar: i32,
    baz: i32,
};

fn fooCallback(foo: *Foo) void {
    out.print("Yo! I was just called. bar = {}, baz = {}.", .{foo.bar, foo.baz}) catch unreachable;
}

pub fn main() !void {
    var foo = Foo {
        .bar = 30,
        .baz = 50,
    };

    const entry = CallbackEntry {
        .target = @ptrCast(*CallbackEntry.Context, &foo),
        .func = @ptrCast(fn(*CallbackEntry.Context)void, fooCallback), // [1]
    };

    entry.func(entry.target); // [2]
}
```
**How I solved:**
[1] - I also noticed that when I pass a function identifier in Zig, I'm passing its pointer, so &fooCallback was actually passing a "pointer to a function pointer" which is not what I wanted, so I changed to fooCallback and it just works.
[2] - I removed the unnecessary ".*", I just figured it wasn't necessary, I didn't know Zig would dereference a function pointer automatically (I'm coming from C), I'm still a newbie in Zig, my bad.

It's kinda curious how this compiler "bug" prevented me from doing something really wrong.

**Comment by daurnimator** [comment]
> [2] - I removed the unnecessary ".*", I just figured it wasn't necessary, I didn't know Zig would dereference a function pointer automatically (I'm coming from C), I'm still a newbie in Zig, my bad.

This is going to be fixed up as part of #1717

**Comment by davi2205** [comment]
> > [2] - I removed the unnecessary ".*", I just figured it wasn't necessary, I didn't know Zig would dereference a function pointer automatically (I'm coming from C), I'm still a newbie in Zig, my bad.
> 
> This is going to be fixed up as part of #1717

Cool! 

I like the new function as expression thing but I've got mixed feelings about removing the status quo* function declaration syntax.

* as of today
